components/sceneutil/workqueue.cpp

```diff
@@ -19,10 +19,8 @@ void WorkItem::waitTillDone()
 
 void WorkItem::signalDone()
 {
-    {
-        OpenThreads::ScopedLock<OpenThreads::Mutex> lock(mMutex);
-        mDone.exchange(1);
-    }
+    OpenThreads::ScopedLock<OpenThreads::Mutex> lock(mMutex);
+    mDone.exchange(1);
```

Comment:
I'm not sure this is necessary. I don't know the exact details of the code, but it's a more preferable paradigm to release the lock after setting the checked condition and before signalling the actual `Condition` variable. Like this:
```
// Thread 1
OpenThreads::ScopedLock<OpenThreads::Mutex> lock(mMutex);
while(!mDone)
    mCondition.wait(lock);

// Thread 2
{
    OpenThreads::ScopedLock<OpenThreads::Mutex> lock(mMutex);
    mDone = 1;
}
mCondition.broadcast();
```
This is completely safe (no unsynchronized access to `mDone`) and avoids extraneous wakeups from `broadcast()` causing Thread 1 to wakeup and try to retake the lock while Thread 2 still has the lock, causing Thread 1 to immediately go back to sleep, while Thread 2 then continues on and finally releases the lock. The OS's scheduler really doesn't like threads constantly waking up and going back to sleep like that, especially when it actually has something to do but couldn't because of that microsecond of contention.

Reply:
I did not invent anything - in all other cases in the engine we do the broadcast under lock.

Reply:
I didn't say you did invent anything. All I'm saying is it's preferable, when possible, to not call `broadcast()` or `notify()` with the lock held to avoid scheduling conflicts. And in the case here, it does not seem necessary to call `broadcast()` with the lock held, as long as the lock was held when `mDone` was set.